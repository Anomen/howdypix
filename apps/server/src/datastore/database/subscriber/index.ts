export * from "./updateStoreOnPhotoSave";
export * from "./updateStoreOnAlbumSave";
export * from "./updateStoreOnSearchSave";
export * from "./updateStoreOnSearchResultSave";
