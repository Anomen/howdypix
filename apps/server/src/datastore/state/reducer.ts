import { Reducer } from "redux";
import { uniqWith, isEqual } from "lodash";

type StateSearch = {
  lastUpdatedPhoto: Date | null;
  album: string;
  source: string;
};

type StateAlbum = {
  lastUpdatedPhoto: Date | null;
};

type StateSearchAlbumRelation = {
  searchId: number;
  albumId: number;
};

export type State = {
  searches: {
    [id: number]: StateSearch;
  };
  albums: {
    [id: number]: StateAlbum;
  };
  searchAlbumRelation: StateSearchAlbumRelation[];
};

export type Action =
  | {
      type: "RESET_STORE";
      data: State;
    }
  | {
      type: "UPDATE_SEARCH_LAST_UPDATED_PHOTO";
      searchId: number;
      data: {
        lastUpdatedPhoto: Date;
      };
    }
  | {
      type: "STORE_NEW_SEARCH";
      searchId: number;
      data: StateSearch;
    }
  | {
      type: "UPDATE_ALBUM_LAST_UPDATED_PHOTO";
      albumId: number;
      data: {
        lastUpdatedPhoto: Date;
      };
    }
  | {
      type: "STORE_NEW_ALBUM";
      albumId: number;
      data: StateAlbum;
    }
  | {
      type: "STORE_NEW_SEARCH_ALBUM_RELATION";
      albumId: number;
      searchId: number;
    };

// ========================================================================
// Reducer
// ========================================================================
export const reducer: Reducer<State, Action> = (
  state = { searches: {}, albums: {}, searchAlbumRelation: [] },
  action
) => {
  // console.log(action);
  switch (action.type) {
    case "RESET_STORE":
      return action.data;

    case "UPDATE_SEARCH_LAST_UPDATED_PHOTO": {
      const lastUpdatedPhoto =
        state.searches[action.searchId]?.lastUpdatedPhoto;

      if (
        (lastUpdatedPhoto && action.data.lastUpdatedPhoto > lastUpdatedPhoto) ||
        lastUpdatedPhoto === null
      ) {
        return {
          ...state,
          searches: {
            ...state.searches,
            [action.searchId]: {
              ...state.searches[action.searchId],
              lastUpdatedPhoto: new Date(action.data.lastUpdatedPhoto),
            },
          },
        };
      }

      return state;
    }

    case "UPDATE_ALBUM_LAST_UPDATED_PHOTO": {
      const lastUpdatedPhoto = state.albums[action.albumId]?.lastUpdatedPhoto;

      if (
        (lastUpdatedPhoto && action.data.lastUpdatedPhoto > lastUpdatedPhoto) ||
        lastUpdatedPhoto === null
      ) {
        return {
          ...state,
          albums: {
            ...state.albums,
            [action.albumId]: {
              lastUpdatedPhoto: new Date(action.data.lastUpdatedPhoto),
            },
          },
        };
      }

      return state;
    }

    case "STORE_NEW_ALBUM":
      return {
        ...state,
        albums: {
          ...state.albums,
          [action.albumId]: {
            ...action.data,
            lastUpdatedPhoto: action.data.lastUpdatedPhoto
              ? new Date(action.data.lastUpdatedPhoto)
              : null,
          },
        },
      };

    case "STORE_NEW_SEARCH":
      return {
        ...state,
        searches: {
          ...state.searches,
          [action.searchId]: {
            ...action.data,
            lastUpdatedPhoto: action.data.lastUpdatedPhoto
              ? new Date(action.data.lastUpdatedPhoto)
              : null,
          },
        },
        searchAlbumRelation: uniqWith(
          [
            ...state.searchAlbumRelation,
            {
              searchId: action.searchId,
              albumId: action.data.album,
            },
          ],
          isEqual
        ),
      };

    case "STORE_NEW_SEARCH_ALBUM_RELATION":
      return {
        ...state,
        searchAlbumRelation: [
          ...state.searchAlbumRelation,
          ...(state.searchAlbumRelation.find(
            (relation) =>
              relation.albumId === action.albumId &&
              relation.searchId === action.searchId
          )
            ? [{ albumId: action.albumId, searchId: action.searchId }]
            : []),
        ],
      };

    default:
      return state;
  }
};
