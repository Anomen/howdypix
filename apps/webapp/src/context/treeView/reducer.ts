import { uniq, uniqBy } from "lodash";
import { QueryLazyOptions } from "@apollo/react-hooks";
import {
  GetTreeAlbums,
  GetTreeQuery,
  GetTreeQueryVariables,
  GetTreeSources,
} from "../../__generated__/schema-types";
import {
  AlbumWithNodeId,
  SourceWithNodeId,
  TreeItem,
  TreeItemWithParent,
} from "./types";

// ========================================================================
// Utility functions
// ========================================================================
const selectParents = (
  data: GetTreeQuery["getTree"]["albums"] | null | undefined,
  dir: string
): { [key: string]: boolean } => {
  let ret: { [key: string]: boolean } = {};

  if (!data) {
    ret[dir] = true;
  } else {
    data.forEach((d): void => {
      if (d.dir === dir) {
        ret[d.dir] = true;

        if (d.parentDir) {
          ret = {
            ...ret,
            ...selectParents(data, d.parentDir),
          };
        }
      }
    });
  }

  return ret;
};

const calculateExpandedNodeIds = (
  data: AlbumWithNodeId[],
  selectedItems: { [name: string]: boolean }
): string[] => {
  const ret: string[] = [];

  data.forEach((d): void => {
    if (d.dir && selectedItems[d.dir]) {
      ret.push(d.nodeId);
    }
    if (d.parentDir === "" && selectedItems[d.source]) {
      ret.push(d.source);
    }
  });

  return uniq(ret);
};

// ========================================================================
// Types
// ========================================================================
export type State = {
  fetchedAlbums: AlbumWithNodeId[];
  fetchedSources: SourceWithNodeId[];
  visibleLeaves: {
    [key: string]: boolean;
  };
  expandedNodeIds: string[];
};

export type Actions =
  | {
      type: "DATA_FETCHED";
      albums: GetTreeAlbums[];
      sources: GetTreeSources[];
      variables: {
        album: string;
        source: string;
      };
    }
  | {
      type: "DISPLAY_LEAF";
      album: string;
      source: string;
    }
  | {
      type: "UPDATE_EXPANDED_NODE_IDS";
    }
  | (TreeItem & {
      type: "EXPAND";
      fetchTree: (
        options?: QueryLazyOptions<GetTreeQueryVariables> | undefined
      ) => void;
    })
  | (TreeItemWithParent & {
      type: "COLLAPSE";
    })
  | (TreeItemWithParent & {
      type: "TOGGLE";
      fetchTree: (
        options?: QueryLazyOptions<GetTreeQueryVariables> | undefined
      ) => void;
    });

// ========================================================================
// Reducer
// ========================================================================
export const reducer = (state: State, action: Actions): State => {
  let newState: State = state;

  switch (action.type) {
    case "DATA_FETCHED":
      newState = {
        ...newState,
        fetchedSources: uniqBy(
          [...(newState.fetchedSources ?? []), ...action.sources],
          "name"
        ).map(
          (source): SourceWithNodeId => ({
            ...source,
            nodeId: source.name,
          })
        ),
        fetchedAlbums: uniqBy(
          [...(newState.fetchedAlbums ?? []), ...action.albums],
          "dir"
        ).map((album) => ({
          ...album,
          nodeId: album.source + album.dir,
        })),
      };

      return reducer(newState, {
        type: "DISPLAY_LEAF",
        source: action.variables.source,
        album: action.variables.album,
      });

    case "DISPLAY_LEAF":
      return reducer(
        {
          ...newState,
          visibleLeaves: {
            ...newState.visibleLeaves,
            ...(action.album !== ""
              ? selectParents(newState.fetchedAlbums, action.album)
              : {}),
            [action.source]: true,
          },
        },
        { type: "UPDATE_EXPANDED_NODE_IDS" }
      );

    case "UPDATE_EXPANDED_NODE_IDS":
      return {
        ...newState,
        expandedNodeIds: calculateExpandedNodeIds(
          newState.fetchedAlbums,
          newState.visibleLeaves
        ),
      };

    case "EXPAND":
      if (action.album === null) {
        if (newState.visibleLeaves[action.source] === undefined) {
          action.fetchTree({
            variables: {
              album: "",
              source: action.source,
            },
          });
        }
      } else if (newState.visibleLeaves[action.album] === undefined) {
        action.fetchTree({
          variables: {
            album: action.album ?? "",
            source: action.source,
          },
        });
      }

      // different than original
      return reducer(
        reducer(newState, {
          type: "DISPLAY_LEAF",
          source: action.source,
          album: action.album ?? "",
        }),
        { type: "UPDATE_EXPANDED_NODE_IDS" }
      );

    case "COLLAPSE":
      return reducer(
        {
          ...newState,
          visibleLeaves: {
            ...newState.visibleLeaves,
            ...(action.album !== null
              ? { [action.album]: false }
              : { [action.source]: false }),
            ...(action.parent ? { [action.parent]: true } : {}),
          },
        },
        { type: "UPDATE_EXPANDED_NODE_IDS" }
      );

    case "TOGGLE":
      if (action.album !== null) {
        if (newState.visibleLeaves[action.album]) {
          return reducer(newState, {
            ...action,
            type: "COLLAPSE",
          });
        }

        return reducer(newState, {
          ...action,
          type: "EXPAND",
        });
      }

      if (newState.visibleLeaves[action.source]) {
        return reducer(newState, {
          ...action,
          type: "COLLAPSE",
        });
      }

      return reducer(newState, {
        ...action,
        type: "EXPAND",
      });

    default:
      return newState;
  }
};
