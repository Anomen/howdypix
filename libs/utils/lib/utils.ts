import { Channel, ConsumeMessage, Options } from "amqplib";
import { join, parse } from "path";
import { MessageProcess, QueueName } from "@howdypix/shared-types";
import debug from "debug";

export async function wait(seconds: number) {
  return new Promise(resolve => {
    setTimeout(resolve, seconds * 1000);
  });
}

export function howdypixPathJoin(
  root: string,
  sourceId: string,
  path: string = ""
) {
  return join(root, ".howdypix", sourceId, path);
}

export function generateThumbnailPaths(
  thumbnailsDir: string,
  sourceId: string,
  path: string = ""
): Array<{
  width: number | null;
  height: number | null;
  path: string;
}> {
  const { dir, name } = parse(howdypixPathJoin(thumbnailsDir, sourceId, path));

  return [200, 600].map(size => ({
    width: size,
    height: null,
    path: join(dir, `${name}x${size}.jpg`)
  }));
}

export function generateThumbnailUrls(
  baseUrl: string,
  sourceId: string,
  path: string = ""
): Array<{
  width: number | null;
  height: number | null;
  url: string;
}> {
  const { name } = parse(path);

  return [200, 600].map(size => ({
    width: size,
    height: null,
    url: baseUrl + "/static/" + join(sourceId, `${name}x${size}.jpg`)
  }));
}

export function isHowdypixPath(path: string = ""): boolean {
  return path.match(/\.howdypix/) !== null;
}

export function appDebug(space: string) {
  return debug(`app:${space}`);
}

export function libDebug(space: string) {
  return debug(`lib:${space}`);
}

export async function assertQueue(channel: Channel, name: QueueName) {
  return channel.assertQueue(name);
}

export interface ParsedConsumeMessage<T> extends ConsumeMessage {
  data: T;
}

export async function consume<T>(
  channel: Channel,
  name: QueueName,
  onMessage?: (msg: ParsedConsumeMessage<T> | null) => any,
  options?: Options.Consume
) {
  return channel.consume(
    name,
    msg => {
      if (msg) {
        const data: T = JSON.parse(msg.content.toString());

        libDebug("rabbit:consume")(`${name} ${JSON.stringify(data)}`);

        onMessage && onMessage({ ...msg, data });
      } else {
        onMessage && onMessage(null);
      }
    },
    options
  );
}

export async function sendToQueue<T>(
  channel: Channel,
  queue: QueueName,
  data: T,
  options?: Options.Publish
) {
  libDebug("rabbit:sendToQueue")(`${queue} ${JSON.stringify(data)}`);

  return channel.sendToQueue(queue, Buffer.from(JSON.stringify(data)), options);
}
