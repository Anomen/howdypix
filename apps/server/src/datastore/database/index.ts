import * as entities from "./entity";
import * as subscribers from "./subscriber";

export { entities, subscribers };
