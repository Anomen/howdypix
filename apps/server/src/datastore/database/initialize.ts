import { Connection, EntitySchema, getConnectionManager } from "typeorm";
import { map } from "lodash";
import { AppStore } from "../state";
import * as subscribers from "./subscriber";
import * as entities from "./entity";
import { AppConfig } from "../../lib/config";
import { Events } from "../../lib/eventEmitter";

export const initializeDatabase = (
  dbConfig: AppConfig["database"],
  store: AppStore,
  event: Events
): Connection =>
  getConnectionManager().create({
    type: "sqlite",
    database: dbConfig.path,
    synchronize: true,
    logging: false,
    entities: map(entities, (entity) => (entity as unknown) as EntitySchema),
    subscribers: map(
      subscribers,
      (subscriber) => subscriber(event, store) as Function
    ),
  });
