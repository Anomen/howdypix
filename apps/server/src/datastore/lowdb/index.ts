import FileSync from "lowdb/adapters/FileSync";
import lowdb, { LowdbSync } from "lowdb";
import { State } from "../state/reducer";

export const initializeLowDb = (Adapter = FileSync): LowdbSync<State> => {
  const adapter = new Adapter("db.json");
  return lowdb(adapter) as LowdbSync<State>;
};
