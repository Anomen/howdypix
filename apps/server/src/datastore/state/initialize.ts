import { Connection } from "typeorm";
import { LowdbSync } from "lowdb";
import { Store } from "redux";
import { Action, State } from "./reducer";
import { Album, Search } from "../database/entity";

export type AppStore = Store<State, Action>;

export const initializeStore = async (
  store: AppStore,
  connection: Connection,
  statedb: LowdbSync<State>
): Promise<void> => {
  // ======================================
  // Initialize the data
  // ======================================
  // Search for albums
  await connection
    .getRepository(Album)
    .createQueryBuilder("album")
    .select("album.id", "id")
    .addSelect("MAX(photo.updatedAt)", "lastUpdatedPhoto")
    .leftJoinAndSelect("album.photos", "photo")
    .groupBy("album.id")
    .getRawMany()
    .then((data: { id: Album["id"]; lastUpdatedPhoto: Date | null }[]) =>
      data.forEach(({ id, lastUpdatedPhoto }) => {
        store.dispatch({
          type: "STORE_NEW_ALBUM",
          albumId: id,
          data: {
            lastUpdatedPhoto,
          },
        });
      })
    );

  // Search for saved searches
  await connection
    .getRepository(Search)
    .createQueryBuilder("search")
    .select("search.id", "id")
    .addSelect("search.album", "album")
    .addSelect("search.source", "source")
    .addSelect("MAX(photo.updatedAt)", "lastUpdatedPhoto")
    .leftJoinAndSelect("search.searchResults", "searchResult")
    .leftJoinAndSelect("searchResult.photo", "photo")
    .groupBy("search.id")
    .getRawMany()
    .then(
      (
        data: {
          id: Search["id"];
          album: Search["album"];
          source: Search["source"];
          lastUpdatedPhoto: Date | null;
        }[]
      ) =>
        data.forEach(({ id, album, source, lastUpdatedPhoto }) => {
          store.dispatch({
            type: "STORE_NEW_SEARCH",
            searchId: id,
            data: {
              album,
              source,
              lastUpdatedPhoto,
            },
          });
        })
    );

  // Persist the store
  store.subscribe(() => {
    statedb.setState(store.getState()).write();
  });
};
