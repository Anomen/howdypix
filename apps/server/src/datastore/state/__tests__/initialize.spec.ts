import Memory from "lowdb/adapters/Memory";
import { wait } from "@howdypix/utils";
import { reducer, State } from "../reducer";
import { createAppStore, initializeStore } from "../index";
import { initializeLowDb } from "../../lowdb";
import { initializeDatabase } from "../../database/initialize";
import {
  Album,
  Photo,
  Search,
  SearchResult,
  Source,
} from "../../database/entity";
import {
  generateAlbum,
  generatePhoto,
  generateSearch,
  generateSource,
  mockDate,
} from "../../../lib/testUtils";

describe.skip("state initiliaze", () => {
  const store = createAppStore();
  const statedb = initializeLowDb(Memory);
  const connection = initializeDatabase({ path: ":memory:" }, store);

  beforeEach(async () => {
    if (connection.isConnected) {
      await connection.close();
    }

    await statedb.setState({ albums: {}, searches: {} }).write();
    await connection.connect();
  });

  it("should initialize properly the store", async () => {
    const sourceRepository = connection.getRepository(Source);
    const albumRepository = connection.getRepository(Album);
    const photoRepository = connection.getRepository(Photo);
    const searchRepository = connection.getRepository(Search);
    const searchResultRepository = connection.getRepository(SearchResult);

    jest.spyOn(store, "dispatch");

    const source = await sourceRepository.save(generateSource(1));
    const album = await albumRepository.save(generateAlbum(2, source));

    const photo1 = await photoRepository.save(generatePhoto(3, album));
    await wait(1);
    const photo2 = await photoRepository.save(generatePhoto(4, album));

    const search = await searchRepository.save(
      generateSearch(6, [photo1, photo2])
    );
    await Promise.all(
      search.searchResults.map((searchResult) =>
        searchResultRepository.save(searchResult)
      )
    );

    const utc = new Date().getTimezoneOffset() * 60000;

    // test
    await initializeStore(store, connection, statedb);
    expect(store.getState()).toEqual({
      albums: {
        2: {
          lastUpdatedPhoto: new Date(photo2.updatedAt.getTime() + utc),
        },
      },
      searches: {
        6: {
          album: "album6",
          source: "source6",
          lastUpdatedPhoto: new Date(photo2.updatedAt.getTime() + utc),
        },
      },
    });
  });
});
