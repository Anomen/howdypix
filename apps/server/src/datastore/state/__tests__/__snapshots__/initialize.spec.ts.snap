// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`state initiliaze should initialize properly the store 1`] = `
Object {
  "albums": Object {
    "0": Object {
      "lastUpdatedPhoto": 1970-01-01T00:00:00.002Z,
    },
  },
  "searches": Object {},
}
`;
